templates/CRM/Contact/Form/Search/Builder.js

```diff
@@ -123,7 +149,11 @@
       }
     }
     $.each(CRM.searchBuilder.fieldOptions[field], function(key, option) {
-      var selected = ($.inArray(''+option.key, options) > -1) ? 'selected="selected"' : '';
+      var optionKey = option.key;
+      if ($.inArray(field, CRM.searchBuilder.searchByLabelFields) >= 0) {
+        optionKey = option.value;
+      }
+      var selected = ($.inArray(''+optionKey, options) > -1) ? 'selected="selected"' : '';
       select.append('<option value="' + option.key + '"' + selected + '>' + option.value + '</option>');
```

Comment:
On a closer look I agree with your changes. You are right it won't be appropriate to pass operator for each field on basis of dataType. I tested everything except there is slight miss. It would be 
- select.append('<option value="' + option.key + '"' + selected + '>' + option.value + '</option>');
+ select.append('<option value="' + optionKey + '"' + selected + '>' + option.value + '</option>');

Please update the changes and I'll merge then. Thanks for your work :)